CSG application that takes in a csg tree and prunes away patches, cha…

Jurwen opened the pull request:
…ins, and corners that are not on the CSG shape

Jurwen commented on src/csg.cpp:
```diff
@@ -0,0 +1,174 @@
+    return std::pair(interval, af);
+}
+
+void prune_data(const std::vector<std::array<double, 3>>& mesh_pts,
```
Here, the pruning of the previous meta data happens. Only the patches, chains, and non_manifold_edges_of_vert are changed such that if they are not on the CSG shape, they will be removed. All the other meta data remain unchanged.

qnzhou commented on src/csg.h:
```diff
@@ -0,0 +1,78 @@
+
+#include "implicit_arrangement.h"
+
+///
+///The structure of a csg unit
+///A nested data structure where `operation` takes in a boolean operation, and the elements contain either a function index or later csg unit, where netaives represent the indices, and positives reprsent the index in the CSG tree.
+///e.g. {Intersection, -1, 2} represents an intersection operation between the first function and the second csg unit.
+///
+struct csg_unit{
+    int operation;
+    std::array<int, 2> elements;
+};
```
I thought we decided to take a call back function instead of explicit CSG tree?

```c++
std::function<bool(std::vecrtor<bool>)> keep;
// The function keep takes a list of signs (one per implicit functions) and
// outputs a bool indicating whether the CSG results contains a point with 
// such sign combination.
```

This way, we can hide the specific implementation details used to evaluate the CSG tree.

qnzhou commented on src/csg.h:
```diff
@@ -0,0 +1,78 @@
+
+#ifndef csg_h
+#define csg_h
+#include <fstream>
+#include <string>
+#include <nlohmann/json.hpp>
```
There is no need of including `fstream` and `json.hpp`.

Jurwen commented:
## Update on building a clean API for CSG application

* The library `csg.h` disgards the details of CSG tree structure and instead takes in a lambda function now.
* Updated the pruning of the `shells` such that the topology and patch information they contained is now correct. 
* Updated unit tests so that they cover wider range of scenarios. The exact the configuration is as the following: For each set of example, the bottom has the `shells` printed, and the right-most picture is the implicit arrangement without the boolean operations. The example 2 and 4 shows the case where pre-CSG contains 1 connected components, and post-CSG contains 2. The `shells` correctly reflects this update. 

<img width="866" alt="image" src="https://github.com/duxingyi-charles/Robust-Implicit-Surface-Networks/assets/46691179/8f922bfb-405d-4d32-877c-5a2ffb57968e">

qnzhou commented on tests/test_implicit_networks.cpp:
```diff
@@ -808,3 +810,405 @@ TEST_CASE("material interface on a failed example", "[MI][examples][!shouldfail]
+        std::cout << "corners: " << corners_count << std::endl;
+        std::cout << "cell labels: " << std::endl;
+        // check
+        REQUIRE(patches.size() == 4);
+        REQUIRE(chains.size() == 2);
+        REQUIRE(corners_count == 0);
+        REQUIRE(shells.size() == 2);
```
Is this correct? Is the following visualization corresponding to this test?
<img width="791" alt="image" src="https://github.com/duxingyi-charles/Robust-Implicit-Surface-Networks/assets/3606672/f95350ed-32cf-4815-b555-1c44eec72fdc">

Jurwen commented:
## Update on the output of CSG metadata

* Replaced outputting `shells` with a boolean label `patch_sign_label` for each patch. This boolean label indicates whether the patch has the same orientation with with the geometry. Now, a `true` label represents the normal of the patch's orientation points inside towards the shape, and vice versa. 
* The following configurations correspond to the current unit tests. Notice that the third configuration is slightly changed for a slightly complex patch orientation. The black dots are added for each green patch where they are offset points toward the normal direction.

<img width="1261" alt="image" src="https://github.com/duxingyi-charles/Robust-Implicit-Surface-Networks/assets/46691179/3980b05e-ea01-4876-8079-f4de4f46fddc">

qnzhou reviewed (approved):
Looks good to me! Thank you, Owen!
